cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -181,33 +184,33 @@ public String getFileHash(Path relativePath, String commitHash) throws IOExcepti
       // Find the node representing the exact file path in the tree.
       TreeWalk walk = TreeWalk.forPath(git.getRepository(), realRelativePath.toString(), commit.getTree());
       if (walk == null) {
-        throw new NotFoundException("File not found in Git revision tree.");
+        throw new NotFoundException( "File not found in Git revision tree.");
       }
       return walk.getObjectId(0).getName();
     } finally {
       // Restore the changes to working directory.
-      git.checkout().setName(previousCommit.getName()).call();
+      git.checkout().setName(previousRef).call();
       popStash(stashID);
     }
   }
 
   /**
-   * Initializes the Git repository by cloning remote. This method doesn't re-clone if the remote has already been
-   * cloned earlier.
+   * Initializes the Git repository by cloning remote. If the repository is already cloned, it
+   * resets the git repository to the state in the remote. All local changes will be lost.
    *
    * @return the commit ID of the present HEAD.
    * @throws GitAPIException               when a Git operation fails.
    * @throws IOException                   when file or network I/O fails.
    * @throws AuthenticationConfigException when there is a failure while fetching authentication credentials for Git.
    */
   public String cloneRemote() throws Exception {
+    credentialsProvider.refresh();
     if (git != null) {
-      return resolveHead().getName();
+      return resetToCleanBranch(git.getRepository().getBranch());
```

Comment:
Will we have any check for local changes before reset and getting confirmation from user for losing local changes? There can be more than one user uploading pipelines to git isn't it?

Reply:
No this method will force reset. An additional parameter can be passed to "force" a reset or raise an exception, but there is no use case for it. Presently a new repository manager is created for each push/pull/list operation, so the same repository manager can't have multiple users.
The java doc mentions that the repository manager is not thread safe. The users need to hold a lock on the repository manager till the operation is complete.